// iota-is-sdk/lib/models/schemas/channel-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelLogRequestOptionsSchema = exports.ChannelInfoSearchSchema = exports.ChannelInfoSchema = exports.ChannelAddressSchema = exports.TopicSchema = void 0;
const typebox_1 = require("@sinclair/typebox");
exports.TopicSchema = typebox_1.Type.Object({
    type: typebox_1.Type.String(),
    source: typebox_1.Type.String()
});
exports.ChannelAddressSchema = typebox_1.Type.String({ minLength: 105, maxLength: 105 });
exports.ChannelInfoSchema = typebox_1.Type.Object({
    channelAddress: typebox_1.Type.String({ minLength: 105, maxLength: 105 }),
    authorId: typebox_1.Type.String({ minLength: 50, maxLength: 53 }),
    subscriberIds: typebox_1.Type.Optional(typebox_1.Type.Array(typebox_1.Type.String({ minLength: 50, maxLength: 53 }))),
    topics: typebox_1.Type.Array(exports.TopicSchema),
    created: typebox_1.Type.Optional(typebox_1.Type.String({ format: 'date-time' })),
    latestMessage: typebox_1.Type.Optional(typebox_1.Type.String({ format: 'date-time' }))
});
exports.ChannelInfoSearchSchema = typebox_1.Type.Object({
    authorId: typebox_1.Type.Optional(typebox_1.Type.String()),
    author: typebox_1.Type.Optional(typebox_1.Type.String()),
    topicType: typebox_1.Type.Optional(typebox_1.Type.String()),
    topicSource: typebox_1.Type.Optional(typebox_1.Type.String()),
    created: typebox_1.Type.Optional(typebox_1.Type.String({ format: 'date-time' })),
    latestMessage: typebox_1.Type.Optional(typebox_1.Type.String({ format: 'date-time' })),
    limit: typebox_1.Type.Optional(typebox_1.Type.Number()),
    index: typebox_1.Type.Optional(typebox_1.Type.Number())
});
exports.ChannelLogRequestOptionsSchema = typebox_1.Type.Object({
    ascending: typebox_1.Type.Boolean(),
    limit: typebox_1.Type.Optional(typebox_1.Type.Number()),
    index: typebox_1.Type.Optional(typebox_1.Type.Number()),
    startDate: typebox_1.Type.Optional(typebox_1.Type.String({ format: 'date-time' })),
    endDate: typebox_1.Type.Optional(typebox_1.Type.String({ format: 'date-time' }))
});
