// iota-is-sdk/lib/models/schemas/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentitySchema = exports.IdentityWithoutIdFields = void 0;
const typebox_1 = require("@sinclair/typebox");
const identity_1 = require("./identity");
const verification_bodies_1 = require("./request-response-body/verification-bodies");
exports.IdentityWithoutIdFields = {
    username: typebox_1.Type.Optional(typebox_1.Type.String({ minLength: 3 })),
    registrationDate: typebox_1.Type.Optional(typebox_1.Type.String({ format: 'date-time' })),
    verifiableCredentials: typebox_1.Type.Optional(typebox_1.Type.Array(identity_1.VerifiableCredentialSchema)),
    role: typebox_1.Type.Optional(typebox_1.Type.String()),
    claim: typebox_1.Type.Optional(verification_bodies_1.ClaimSchema),
    isPrivate: typebox_1.Type.Optional(typebox_1.Type.Boolean()),
    isServerIdentity: typebox_1.Type.Optional(typebox_1.Type.Boolean())
};
exports.IdentitySchema = typebox_1.Type.Object(Object.assign({ id: typebox_1.Type.String({ minLength: 50, maxLength: 53 }), publicKey: typebox_1.Type.String({ minLength: 10 }) }, exports.IdentityWithoutIdFields));
